### PR TITLE
[sinks] Create healthchecker and surface indefinite errors

### PR DESCRIPTION
**Pull request by cjubb39**
Modeled off of the source healthchecker, create a new sink healthchecker that reports indefinite errors to a persistent collection.  At the moment, sinks only produce indefinite errors so this is relatively straighforward! We take the error, give it a timestamp at the current time (as determined by the storage controller) and then write it out to a persist shard.

A couple design decisions to note:
- @bkirwi and I have talked about removing the initial snapshotting from the healthchecker.  lmk if you'd like me to roll that into this change
- I used an `Arc<Mutex<Healthchecker>>` within the kafka sink instead of something like `Rc<RefCell<Healthchecker>>`.  While we don't strictly _need_ the `Arc<Mutex>` (since the healthchecker is only used on a single thread), it was tricky to get both the borrow checker and clippy to allow compilation at the same time.  There shouldn't be any lock contention and the code is more future-proofed since we don't have to worry about the single-thread restriction.
- It's becoming increasingly clear that we'll want to reorganize the `KafkaSinkState` struct at some point -- there are a decent number of gymnastics to appease the borrow checker because different subfields have mutable borrows at overlapping points.  All that said, I don't think rolling into this PR is the place to do this.

### Motivation
https://github.com/MaterializeInc/materialize/issues/15163
https://github.com/MaterializeInc/materialize/issues/15164

### Tips for reviewer
I've currently disabled the use of the introspection collection (as the source's is also currently disabled) but I did do a test run with it enabled: https://buildkite.com/materialize/tests/builds/43940

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/storage/src/sink/kafka.rs [hunk 1]
```diff
@@ -348,43 +416,58 @@ struct ProgressInitState {
...
+    healthchecker: Arc<Mutex<Option<Healthchecker>>>,
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 enum KafkaSinkStateEnum {
     // Initialize ourselves as a transactional producer with Kafka
     // Note that this only runs once across all workers - it should only execute
     // for the worker that will actually be publishing to kafka
-    Init(Option<ProgressInitState>),
```
Comment (cjubb39) [inline src/storage/src/sink/kafka.rs:372]: this is unrelated cleanup that we can do now that we are always in "reuse topic" mode!

File: src/storage/src/sink/kafka.rs [hunk 2]
```diff
@@ -334,43 +402,58 @@ struct ProgressInitState {
...
 
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 struct ProgressRunningState {
     topic: String,
     key: String,
     gate_ts: Rc<Cell<Option<Timestamp>>>,
+    // Using Arc<Mutex> because clippy complains about holding a RefCell reference across await points if we use
+    // an `Rc<RefCell>`.
+    healthchecker: Arc<Mutex<Option<Healthchecker>>>,
```
Comment (bkirwi) [inline src/storage/src/sink/kafka.rs:426]: Does this need to live in the progress state? It would be nice if we could restrict this to only managing progress-topic-relevant stuff.

Feels like we might be able to share it the same way we share state like `shutdown_flush` or `metrics` in `produce_to_kafka`?

Comment (cjubb39) [reply src/storage/src/sink/kafka.rs:426]: hmm -- I put it here, because, like the `gate_ts`, it's not something we're able to construct until just before we enter the running state -- and they're things that we definitely want to construct (well, try to anyway, given the Option) before moving into the running state

That being said, you do make a good point.  What do you think of pushing to until we tackle the 
> It's becoming increasingly clear that we'll want to reorganize the KafkaSinkState struct at some point

I mentioned in the description?

File: src/storage/src/types/sinks.rs [hunk 3]
```diff
@@ -29,16 +30,17 @@ include!(concat!(env!("OUT_DIR"), "/mz_storage.types.sinks.rs"));
 
 /// A sink for updates to a relational collection.
 #[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
-pub struct StorageSinkDesc<S = (), T = mz_repr::Timestamp> {
+pub struct StorageSinkDesc<S = (), I = GlobalId, T = mz_repr::Timestamp> {
```
[the remark this answers is not in the document]
Comment (cjubb39) [reply src/storage/src/types/sinks.rs:33]: yeah -- I'll try to whip up a separate PR to do some trait stuff

File: src/storage/src/sink/kafka.rs (hunk 2 above)
Comment (bkirwi) [reply src/storage/src/sink/kafka.rs:426]: > it's not something we're able to construct until just before we enter the running state.

why's this? feels like we'd want to initialize it early in the lifecycle, so we can promptly report the `starting` state.

File: src/storage/src/types/sinks.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bkirwi) [reply src/storage/src/types/sinks.rs:33]: You might be able to get away without any trait machinery; something like:

```
pub struct StorageSinkDesc<S, T = mz_repr::Timestamp> {
  some_field: S,
  ...
}
pub struct ExportStuff {
  id: GlobalId
}
pub struct ExportDescription<T = mz_repr::Timestamp> {
  pub sink: StorageSinkDesc<ExportStuff, T>,
  ...
}
```

Or whatever! But I haven't worked through the details.

File: src/storage/src/sink/kafka.rs [hunk 4]
```diff
@@ -879,19 +907,43 @@ impl KafkaSinkState {
...
             if min_frontier > self.latest_progress_ts {
                 // record the write frontier in the progress topic.
-                if let Some(progress_state) = self.sink_state.unwrap_running() {
-                    self.retry_on_txn_error(|p| p.begin_transaction()).await;
+                let () = match self
+                    .producer
+                    .retry_on_txn_error(|p| p.begin_transaction())
+                    .await
+                {
+                    Ok(()) => (),
```
Comment (bkirwi) [inline src/storage/src/sink/kafka.rs:915]: This pattern gets repeated a ton in this branch... how would you feel about a replace with an `if let Err(err)`, or factoring it out to a helper?

File: src/storage/src/sink/kafka.rs [hunk 5]
```diff
@@ -1058,21 +1113,57 @@ where
     builder.build(move |_capabilities| async move {
```
Comment (bkirwi) [inline src/storage/src/sink/kafka.rs:1113]: Re: our discussion in chat: IIUC we could just initialize the healthchecker here, without needing to inspect the `sink_state`... it's guaranteed to be `Init` here. It looks like with the new setup we don't need the state machine at all, which is great.

Something for a future PR, though.

File: src/storage/src/sink/healthcheck.rs [hunk 6]
```diff
@@ -0,0 +1,715 @@
...
+    /// having to assume that the `upper` is a single `u64`.
+    async fn bootstrap_state(
+        &mut self,
+        mut read_handle: ReadHandle<SourceData, (), Timestamp, i64>,
+        upper: &Antichain<Timestamp>,
+    ) {
+        let since = read_handle.since().clone();
+        trace!("Bootstrapping state as of {:?}!", since);
+        // Ensure the collection is readable at `since`
+        if PartialOrder::less_than(&since, &self.upper) {
```
Comment (bkirwi) [inline src/storage/src/sink/healthcheck.rs]: This still has the bug that I proposed a fix to in that other branch, before I decided I wasn't going to use this implementation in the source. (`self.upper` is always `minimum` here, so this branch is never taken, and we always ignore the content of the state.)

I'm okay merging this because this code is never invoked in this branch, but we should make sure to replace this implementation by the time we enable things.

File: src/storage/src/sink/healthcheck.rs [hunk 7]
```diff
@@ -147,7 +140,8 @@ impl Healthchecker {
                         // Sync to the new upper, go to the loop again
                         self.sync(&actual_upper.0).await;
```
Comment (bkirwi) [inline src/storage/src/sink/healthcheck.rs:141]: FWIW I think we can plausibly now just set `self.upper = actual_upper` here directly and drop the `sync` /  `listener` stuff. Should further reduce the load on persist, and I'm not sure we're getting a ton out of it... loading a lot of data, and the CAS loop will already handle a stale upper correctly.

File: src/storage/src/sink/kafka.rs (hunk 4 above)
Comment (bkirwi) [reply src/storage/src/sink/kafka.rs:915]: (I appreciate this was introduced to deal with my previous comment... 😅)

Comment (cjubb39) [reply src/storage/src/sink/kafka.rs:915]: I'm not a big fan of `if let Err ...` for future-proofing reasons and I think a helper basically puts us back to where we were before with the healthchecker being passed through.  I can do this though (with some futures combinators pre-await) -- but it requires adding an extra two lines for type annotations / because the NeverType isn't stabilized yet
```
               let () = s
                    .producer
                    .retry_on_txn_error(|p| p.init_transactions())
                    .or_else(|err| async {
                        Healthchecker::report_stall_and_panic(healthchecker.as_mut(), err).await
                    })
                    .await
                    .map_err(|()|())
                    .expect("unreachable: can remove once ! stabilizes");
```

It is a little junk (but junk that we should soon (tm) be able to get rid of!) -- but I think it's readable and more future proof than what's currently there!

What do you think?

Comment (bkirwi) [reply src/storage/src/sink/kafka.rs:915]: Mmm, I see what you're after there, but still feels a bit heavy to me.

Not worth burning too many cycles on probably, so happy to leave alone for now; thanks for considering!
